### ISSUE TITLE
204 causes cohtt-async to hang

### ISSUE THREAD

**Issue by sgrove** [mirage/ocaml-cohttp#666]
It looks like all of the informational responses mentioned here https://github.com/mirage/ocaml-cohttp/pull/355 cause cohttp-async client to hang until the server times out the connection:

```reason
let testCall = (code: int) => {
  let start = Unix.time();

  Cohttp_async.Client.get(
    Uri.of_string(Format.sprintf("https://httpstat.us/%d", code)),
  )
  >>= (
    ((resp, body)) => {
      let status =
        resp |> Cohttp.Response.status |> Cohttp.Code.string_of_status;
      let debug =
        Format.asprintf(
          "Status: %s\nFull response:\n%a\n",
          status,
          Cohttp.Response.pp_hum,
          resp,
        );

      Format.printf("Cohttp result:\n%s\n", debug);
      Cohttp_async.Body.to_string(body);
    }
  )
  >>| (
    _body => {
      let duration = Unix.time() -. start;
      Format.printf("Request took: %fs\n", duration);
    }
  );
};

/* All these should complete instantly */
let testWorking = () => {
  testCall(200)
  >>= (_ => testCall(201))
  >>= (_ => testCall(300))
  >>= (_ => testCall(301))
  >>= (_ => testCall(400))
  >>= (_ => testCall(401))
  >>= (_ => testCall(500));
};

/* Each of these will hang */
let testFailing = () => {
  testCall(204);
  testCall(304);
  testCall(100);
  testCall(101);
  testCall(102);
};

let command =
  Async.Command.async_spec(
    ~summary="",
    Command.Spec.empty,
    () => {
      testWorking();
      testFailing();
    },
  );

Command.run(command);
```

In the cases where responses include a `content-length` header, the client works as expected:
```
// Has a content-length: 0 header so the client is fine:
testCall(200);
Cohttp result:
Status: 200 OK
Full response:
((encoding (Fixed 0))
 (headers
  ((access-control-allow-origin *) (cache-control private) (content-length 0)
   (date "Sun, 05 May 2019 22:05:48 GMT") (server Microsoft-IIS/10.0)
   (set-cookie
    "ARRAffinity=580ecdc04582a8c27680686941822edec8921fe0a4244b99f898ac1d3a4de89a;Path=/;HttpOnly;Domain=httpstat.us")
   (x-aspnet-version 4.0.30319) (x-aspnetmvc-version 5.1)
   (x-powered-by ASP.NET)))
 (version HTTP_1_1) (status OK) (flush false))

Request took: 0.000000s
- : unit = ()
```

But it seems all of the informational responses (correctly) do not include a `content-length` header, and this causes the client to hang:

```
// Note the lack of content-length header, and that the request took 129 seconds
testCall(204);
Cohttp result:
Status: 204 No Content
Full response:
((encoding Unknown)
 (headers
  ((access-control-allow-origin *) (cache-control private)
   (date "Sun, 05 May 2019 22:07:18 GMT") (server Microsoft-IIS/10.0)
   (set-cookie
    "ARRAffinity=580ecdc04582a8c27680686941822edec8921fe0a4244b99f898ac1d3a4de89a;Path=/;HttpOnly;Domain=httpstat.us")
   (x-aspnet-version 4.0.30319) (x-aspnetmvc-version 5.1)
   (x-powered-by ASP.NET)))
 (version HTTP_1_1) (status No_content) (flush false))

Request took: 129.000000s
- : unit = ()
```

Presumably the client is waiting to read more input from the server, until the server times out the connection.

I think this has been the case for awhile - one of the services we were interacting with returns lots of 204 responses, but previously they included a `content-length` header. Once they removed that (presumably to be "more correct") we started hitting this issue, and it seems to be reproducible across several services.

**Comment by emillon** [comment]
Hi @sgrove!

> It looks like all of the informational responses mentioned here #355 cause cohttp-async client to hang until the server times out the connection

This is exactly what's happening, in the absence of `Content-Length` it hits this part:

https://github.com/mirage/ocaml-cohttp/blob/f349fa767908a41c609fc3a87c2f0c69c3f180fb/cohttp/src/transfer_io.ml#L108-L114

I pushed a fix in #671, let me know if that fixes the issue with the external services you're hitting.

Thanks!

**Comment by emillon** [comment]
(it only fixes 204 & 304 though, I suppose that 100, 101 and 102 in the `testFailing` list need special support)
